### PR TITLE
Add inventory chunks trimming logic based on newline delimiters

### PR DESCRIPTION
**Pull request by bassemwanis**
*Description of changes:*
Add inventory chunks trimming logic based on newline delimiters

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: source/refreezer/application/chunking/trim.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+def trim_inventory_chunk(
+    first_chunk: bool, last_chunk: bool, max_record_size: int, chunk: bytes
+) -> bytes:
+    if not first_chunk:
+        """
+        Remove all bytes from the beginning of the chunk array up to and
+        including the last occurrence of the newline character \n within
+        the first max_record_size bytes of the chunk array
+        """
+        chunk = chunk[chunk.rindex(b"\n", 0, max_record_size) + 1 :]
```
Comment (dbb-aws-only) [inline source/refreezer/application/chunking/trim.py:16]: Is it possible there will be no \n in a chunk?  The rindex will return a "Substring not found" ValueError in that case.

**Review by dbb-aws-only — DISMISSED** [review]
As long as we're ok with not trapping a ValueError in the rindex on lines 16 and 23 - this code looks great

File: source/refreezer/application/chunking/trim.py (hunk 1 above)
Comment (bassemwanis) [reply source/refreezer/application/chunking/trim.py:16]: Right, I was thinking that the chunk size should always be greater than the `max_record_size` and should have at least one `\n`. Also since a `ValueError` will be raised, can the callee logic handle it!

Comment (kamyarz-aws) [inline source/refreezer/application/chunking/trim.py:16]: docstring best practice is to use it to explain what a function does. In this case the best practice is to simply use multiple lines of comments.
image

File: source/tests/unit/application/chunking/test_trim_inventory_chunk.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+from refreezer.application.chunking.trim import trim_inventory_chunk
+
+
+def test_trim_inventory_first_chunk() -> None:
+    max_record_size = 10
+    chunk = b"abcdef\nghi\nminopqrnq\nstqauvwx\nyz"
+
+    expected_result = b"abcdef\nghi\nminopqrnq\nstqauvwx"
+    result = trim_inventory_chunk(True, False, max_record_size, chunk)
+    assert expected_result == result
```
[the remark this answers is not in the document]
Comment (bassemwanis) [reply source/tests/unit/application/chunking/test_trim_inventory_chunk.py]: Done!

File: source/refreezer/application/chunking/trim.py (hunk 1 above)
Comment (bassemwanis) [reply source/refreezer/application/chunking/trim.py]: Done!

**Review by dbb-aws-only — DISMISSED** [review]

**Review by kamyarz-aws — DISMISSED** [review]

File: source/refreezer/application/chunking/trim.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+
+def trim_inventory_chunk(
+    first_chunk: bool, last_chunk: bool, max_record_size: int, chunk: bytes
+) -> bytes:
+    if not first_chunk:
+        # Remove all bytes from the beginning of the chunk array up to and
+        # including the last occurrence of the newline character \n within
+        # the first max_record_size bytes of the chunk array
+        chunk = chunk[chunk.rindex(b"\n", 0, max_record_size) + 1 :]
```
Comment (schana) [inline source/refreezer/application/chunking/trim.py:14]: We should add edge-case testing for what happens when the newline ends up on the boundaries of the max_record_size as well as chunks.

**Review by schana — CHANGES_REQUESTED** [review]

File: source/refreezer/application/chunking/trim.py (hunk 3 above)
Comment (bassemwanis) [reply source/refreezer/application/chunking/trim.py:14]: Done! Also I have added a new test to chunk an input byte array, trim each chunk, and then reassemble the chunks back into a single byte array.

**Review by dbb-aws-only — APPROVED** [review]
The testing for this code is excellent - I love the corner cases that you've covered.  I think it's ready to merge now.

File: source/refreezer/application/chunking/trim.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+
+def trim_inventory_chunk(
+    first_chunk: bool, last_chunk: bool, max_record_size: int, chunk: bytes
+) -> bytes:
+    if not first_chunk:
+        # Remove all bytes from the beginning of the chunk array up to
+        # (not including) the last occurrence of the newline character \n
+        # within the first max_record_size bytes of the chunk array
+        chunk = chunk[chunk.rindex(b"\n", 0, max_record_size) :]
```
Comment (dbb-aws-only) [inline source/refreezer/application/chunking/trim.py:14]: I know this doesn't belong in a code review, but the Robot Smile  `:]` makes me happy.

**Review by simonkrol — APPROVED** [review]

**Review by schana — APPROVED** [review]
